squeeze: allow squeezing multiple dimensions at once

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91602
* __->__ #89017

Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze((0, 1))
```
(assuming x has at least 2 dimensions)

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

peterbell10 committed:
squeeze: allow squeezing multiple dimensions at once

Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze(0, 1)
```
(assuming x has at least 2 dimensions)

[ghstack-poisoned]

peterbell10 committed:
Update on "squeeze: allow squeezing multiple dimensions at once"


Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze(0, 1)
```
(assuming x has at least 2 dimensions)

[ghstack-poisoned]

peterbell10 committed:
Update on "squeeze: allow squeezing multiple dimensions at once"


Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze(0, 1)
```
(assuming x has at least 2 dimensions)

[ghstack-poisoned]

peterbell10 committed:
Update on "squeeze: allow squeezing multiple dimensions at once"


Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze(0, 1)
```
(assuming x has at least 2 dimensions)

[ghstack-poisoned]

peterbell10 committed:
Update on "squeeze: allow squeezing multiple dimensions at once"


Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze(0, 1)
```
(assuming x has at least 2 dimensions)

[ghstack-poisoned]

peterbell10 committed:
Update on "squeeze: allow squeezing multiple dimensions at once"


Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze((0, 1))
```
(assuming x has at least 2 dimensions)

[ghstack-poisoned]

albanD commented:
@mruberry can you take care of this one?

peterbell10 committed:
Rebase and fix merge conflicts on "squeeze: allow squeezing multiple dimensions at once"


Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze((0, 1))
```
(assuming x has at least 2 dimensions)

[ghstack-poisoned]

peterbell10 committed:
Update on "squeeze: allow squeezing multiple dimensions at once"


Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze((0, 1))
```
(assuming x has at least 2 dimensions)

[ghstack-poisoned]

peterbell10 commented:
@mruberry ping

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5002,6 +5002,17 @@
   device_guard: False
 
+
+- func: squeeze.dims(Tensor(a) self, int[] dim) -> Tensor(a)
```
What is the reason for not re-using the `squeeze.dim` overload above with `int[1]` as the new argument?
Just avoid BC issues?

peterbell10 replied:
I tried in #91603 but it breaks BC so I think we need to keep this.

albanD replied:
Ho sad, ok !

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1506,6 +1506,12 @@
       self: grad.unsqueeze(dim)
 
+- name: squeeze.dims(Tensor(a) self, int[] dim) -> Tensor(a)
+  dispatch:
```
No need for dispatch if you only have the Default entry

albanD commented on torch/_inductor/lowering.py:
```diff
@@ -511,16 +512,17 @@ def squeeze(x, dim=None):
-            return view(x, new_shape)
 
+    dim = canonicalize_dims(len(x.get_size()), dim)
+    dims = set((dim,) if not isinstance(dim, tuple) else dim)
```
Why a set and not a tuple here?

peterbell10 replied:
The `d in dims` check below should be faster with a `set`. Note the whole `if` expression is being used to construct the set, in case that's confusing.

albanD commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -3169,6 +3186,34 @@ Tensor squeeze_quantized(const Tensor& self, int64_t dim) {
+  return result;
+}
+
+Tensor squeeze_multiple(const Tensor& self, IntArrayRef dims) {
```
Why is this not the `squeeze()` function above?

peterbell10 replied:
I've updated the `NestedTensor` and `Quantized` paths to support the multiple dims, so this is no longer necessary.

albanD commented on torch/csrc/autograd/FunctionsManual.h:
```diff
@@ -220,6 +220,10 @@ at::Tensor unsqueeze_to(
     const at::Tensor& self,
     int64_t dim,
     c10::SymIntArrayRef sym_sizes);
+at::Tensor unsqueeze_to(
```
The one above could be removed no?

peterbell10 replied:
Both `squeeze.dim` and `squeeze.dims` call `unsqueeze_to` in their backwards, so we do use both overloads. I've rewritten the `int` version to just call the `int[]` version to reduce duplication though.

albanD reviewed (commented):
Looks good! Mostly curious why we duplicate a lot of functions instead of handling them in a similar way we do for sum?

anijain2305 commented:
Thanks, this will reduce the graph size for many vision models. Currently, we have many squeeze ops as a result of this functions - https://github.com/pytorch/pytorch/blob/master/torch/_decomp/decompositions.py#L1899

If possible, we can get it in this PR. Its totally ok to follow it up in another PR, if it complicates the testing.

ngimel commented:
This won't help graph size for existing models, because squeeze lowering is still for a single dimension.

anijain2305 commented:
> This won't help graph size for existing models, because squeeze lowering is still for a single dimension.

It seems to me that it should? This PR changes inductor lowering as well, that uses a view op with the new shape (by consolidating the dims).

peterbell10 committed:
Update on "squeeze: allow squeezing multiple dimensions at once"


Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze((0, 1))
```
(assuming x has at least 2 dimensions)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "squeeze: allow squeezing multiple dimensions at once"


Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze((0, 1))
```
(assuming x has at least 2 dimensions)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "squeeze: allow squeezing multiple dimensions at once"


Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze((0, 1))
```
(assuming x has at least 2 dimensions)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Fix missing batch rule on "squeeze: allow squeezing multiple dimensions at once"


Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze((0, 1))
```
(assuming x has at least 2 dimensions)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

albanD reviewed (approved):
Ok!

peterbell10 commented:
@pytorchbot merge -f "ROCm timeout is unrelated"

mehtanirav commented:
This PR is causing internal build failures with errors like

```
[2023-01-04T23:03:00.073-08:00] Stderr: fbcode/caffe2/aten/src/ATen/native/nested/NestedTensorMath.cpp:711:20: error: comparison of integers of different signs: 'int64_t' (aka 'long') and 'unsigned long' [-Werror,-Wsign-compare]
[CONTEXT] [2023-01-04T23:03:00.073-08:00]   TORCH_CHECK(ndim > 1 + dims.size(),
[CONTEXT] [2023-01-04T23:03:00.073-08:00]               ~~~~ ^ ~~~~~~~~~~~~~~~
```

mehtanirav commented:
@pytorchbot  revert -m "Internal breakages" -c ghfirst

ngimel commented:
Next PR in the stack will also need to be reverted, why wasn't it reverted automatically?

ngimel commented:
Yes, even though there's no conflicts, #91602 relies on functionality introduced here. Could we have done a quick fix here instead? If unsigned comparison is the only problem, that's an easy forward fix.

peterbell10 committed:
Update on "squeeze: allow squeezing multiple dimensions at once"


Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze((0, 1))
```
(assuming x has at least 2 dimensions)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "squeeze: allow squeezing multiple dimensions at once"


Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze((0, 1))
```
(assuming x has at least 2 dimensions)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Fix merge conflict on "squeeze: allow squeezing multiple dimensions at once"


Ref #70924

This addresses part 1 of the issue, allowing `torch.squeeze` to be
passed a tuple of dimensions. e.g.
```python
x.squeeze(0).squeeze(0)
```
can now be written
```python
x.squeeze((0, 1))
```
(assuming x has at least 2 dimensions)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
